src/wallet/wallet.cpp

```diff
@@ -1582,13 +1588,17 @@ CBlockIndex* CWallet::ScanForWalletTransactions(CBlockIndex* pindexStart, bool f
 
             CBlock block;
             if (ReadBlockFromDisk(block, pindex, Params().GetConsensus())) {
+                LOCK(cs_wallet);
                 for (size_t posInBlock = 0; posInBlock < block.vtx.size(); ++posInBlock) {
                     AddToWalletIfInvolvingMe(block.vtx[posInBlock], pindex, posInBlock, fUpdate);
                 }
             } else {
                 ret = pindex;
             }
-            pindex = chainActive.Next(pindex);
+            {
+                LOCK(cs_main);
+                pindex = chainActive.Next(pindex);
```

Comment:
What if there's a reorg between iterations?

Does anything weird happen if we (for ex): begin scanning, receive a new best block with our tx from p2p, then re-scan here?

Seems to me that we should take a copy of ChainActive, traverse, compare to progress to chainActive.Tip(), and repeat until they match. Something like (just a quick sketch, probably broken):
```c++
CBlockIndex* pindex = pindexStart;
CChain chain;
while (!fAbortRescan) {
    {
        LOCK(cs_main);
        if (pindex == chainActive.Tip()) {
            break;
        }
        chain.SetTip(chainActive.Tip());
    }
    if (pindex) {
        // start where we left off.
        const CBlockIndex* fork = chain.FindFork(pindex);
        if (fork != pindex) {
            // Need to undo the orphans.
        }
        pindex = chain.Next(fork);
    }
    while(!fAbortRescan && pindex) {
        /*
        do stuff
        ...
        pindex = chain.Next(pindex);
        */
    }
    pindex = chain.Tip();
}
```
That almost entirely avoids locking cs_main, while ensuring that we're never jumping chains.

It could also be done with just the pindex and GetAncestor(), to skip the memory overhead of the second chain copy.

Reply:
I'm not sure if this is really a problem. During rescans, IMO it won't hurt to scan an orphaned block nor does it hurt if you do not scan up to the tip in case you have connected new blocks during the time of re-scanning. This, because those new blocks must also have been scanned by the wallet logic (outside of `ScanForWalletTransactions`).

Reply:
If a reorg happens during rescan, we want to continue rescanning using the current chain, not the outdated one.

> This, because those new blocks must also have been scanned by the wallet logic (outside of ScanForWalletTransactions).

Indeed.

Reply:
Maybe break the rescan as soon as possible if the above tip is not in the current chain? Then add here:
```cpp
if (!chainActive.Contains(tip)) break;
```

Reply:
In case of a reorg, wouldn't `chainActive.Next(pindex);` return a `nullptr` leading to stop the scan, which should be totally fine. During the reorg, the blocks of the new chain must have been scanned by the wallet via ConnectBlock.

Reply:
Yes that's true when pindex is not in the chain. But if any later block is not in the chain, why keep rescanning?

Reply:
It should be detected during the next `chainActive.Next(pindex);` which, in case we had a fork between the last `chainActive.Next(pindex);` we would only scan max 1 block... also very unlikely IMO.

Reply:
Suppose the following:
 - `A B C D E F G H I J` - chain to rescan, tip = J
 - `(A) B C D E F G H I J` - scanning block A
 - `A (B) C D E F G H I J` - scanning block B
 - `A (B) C D E F G H K L M` - scanning block B, reorg after H

Should only rescan to H inclusive? Agree it's an edge case.

BTW, if the tip changes `dProgressTip` could be updated?

Reply:
Yea, I think there is a real issue here. I (believe) it could be fixed by doing the AddToWalletIfInvolvingMe inside a cs_main (which is going to be at least mostly required since ReadBlockFromDisk requires cs_main) with a chainActive.Contains() check before doing the AddToWalletIfInvolvingMe. That should put us back to the original state of anything we add to the wallet is on the main chain and, thus, will either be Disconnected then Connected in the callbacks, only Connected, or not included in a callback at all.

Reply:
I still don't fully understand the concerns (please help me).

* If there is a reorg, all blocks of the new chain will be scanned by the wallet via ConnectBlock (no need to rescan again, we only need to rescan the stuff that doesn't come in via Connect/Disconnect block).
* If we scan blocks that are no longer in the main-chain, and we could find a transaction in those blocks that is relevant to our wallet, so be it. The wallet does display such transactions as unconfirmed... same as we would have a reorg outside of the rescan process.

Reply:
Heh, sorry, I didnt fully specify my concern because I thought the above discussion did, but I dont think the above comments are correct. As you point out, I do not believe there is a race where you simply miss transactions. However, there is a race where you may end up marking a transaction as in the wrong block:

A -> B -> ....
    -> C -> ...

Both B and C have the transaction you're interested in.

You go to load block B from a rescan, but it has already been reorg'd and the notificatitons for disconnect(B) and connect(C) have already fired. Thus, you have the transaction marked as in block C, which is correct, but then you finish rescanning and mark the transaction as in block B, which now makes you think the transaction has been conflicted and has negative GetDepthInMainChain.

As I mentioned, this is most simply fixed by adding a "LOCK(cs_main); if (!chainActive.Contains(pindex)) break;" outside the AddToWalletIfInvolvingMe loop.

Reply:
Fixed